PATCH-7: Add VJ TC01

yenleagest opened the pull request:
Hi @nbmhoang,
When you have a chance, please review this pull request. I’ve attached a screenshot as evidence that the new test case has been tested.
![image](https://github.com/user-attachments/assets/69f6c522-a733-458d-930b-4c5b3f7aef50)

catssaymeow committed:
PATCH-7: Add VJ TC01

nbmhoang commented on src/main/java/common/Constants.java:
```diff
@@ -6,10 +6,16 @@ public class Constants {
 
     private static String getTestDataPath() {
-        if (System.getProperty("selenide.baseUrl").contains("agoda")) {
+        String baseUrl = System.getProperty("selenide.baseUrl");
```
Can we get this value from Selenide Configuration.baseUrl?

yenleagest replied:
Updated.

nbmhoang commented on src/main/java/data/models/vj/Ticket.java:
```diff
@@ -0,0 +1,15 @@
+@AllArgsConstructor
+public class Ticket {
+
+    boolean isReturnFlight;
```
Please use enum for this property

yenleagest replied:
Add FlightType enum.

nbmhoang commented on src/main/java/data/models/vj/Ticket.java:
```diff
@@ -0,0 +1,15 @@
+public class Ticket {
+
+    boolean isReturnFlight;
+    private LocalDate returnDate;
```
Why is takeOffDate part of FlightInfo, but returnDate is in Ticket?

yenleagest replied:
As discussed I updated Ticket model.

nbmhoang commented on src/main/java/pages/vj/HomePage.java:
```diff
@@ -0,0 +1,207 @@
+    private final String airportTextBox = "//label[text()='%s']/following-sibling::div/input";
+    private final String airportOption = "//div[contains(@class, 'MuiExpansionPanelDetails')]//div[text()='%s']";
+    private final String passengerValue = "//p[text()='%s']/following::span[contains(@class, 'MuiTypography')][1]";
+    private final String decreaseBtn = passengerValue.concat("/preceding-sibling::button");
```
hmm. Try to find a better way to capture 2 these buttons

yenleagest replied:
Updated.

nbmhoang commented on src/main/java/pages/vj/HomePage.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Step("Search for airport: {airport}")
+    private void searchAirport(String textLocator, String airport) {
```
I suppose this method should be "selectAirport" and  change "enterAirportSearch" to "searchAirport" =))

yenleagest replied:
Updated. Im really bad at naming lol.

nbmhoang commented on src/main/java/pages/vj/HomePage.java:
```diff
@@ -0,0 +1,207 @@
+        String baseUrl = System.getProperty("selenide.baseUrl");
+
+        DateTimeFormatter formatter;
+        if (baseUrl.endsWith("/en"))
```
checking language in current page? please write a method for it.

yenleagest replied:
I did, now we do it in Constants.

nbmhoang commented on src/main/java/pages/vj/HomePage.java:
```diff
@@ -0,0 +1,207 @@
+
+        DateTimeFormatter formatter;
+        if (baseUrl.endsWith("/en"))
+            formatter = DateTimeFormatter.ofPattern("MMMM yyyy", Locale.ENGLISH);
```
put the format pattern in the yaml file also

yenleagest replied:
I moved it to Constants since we have a bunch of formatters. Thought it’d be better to centralize them in one place for easier reuse.

nbmhoang commented on src/main/java/pages/vj/HomePage.java:
```diff
@@ -0,0 +1,207 @@
+    @Step("Get the year and month from the date picker caption")
+    private YearMonth getYearMonthFromDatePicker() {
+        String yearMonth = $(rdrMonthName).shouldBe(visible).shouldNotHave(exactText("")).getText().trim();
+        String baseUrl = System.getProperty("selenide.baseUrl");
```
Same question. Could we get from Configuration.baseUrl or get the title from driver?

yenleagest replied:
Updated.

nbmhoang commented on src/main/java/pages/vj/HomePage.java:
```diff
@@ -0,0 +1,207 @@
+    private final String decreaseBtn = passengerValue.concat("/preceding-sibling::button");
+    private final String increaseBtn = passengerValue.concat("/following-sibling::button");
+    private final String searchBtn = "//span[contains(@class, 'MuiButton')]/span[text()=\"%s\"]";
+    private final String selectedDate = "//div[translate(text(), 't', 'T') = '%s']/following-sibling::div//button[not(contains(@class, 'rdrDayDisabled'))]//span[text()='%s']";
```
Is this element's locator pointing to a selectable day in the calendar, or to the date that has already been selected?

yenleagest replied:
Again, my bad at naming. Updated!

nbmhoang commented on src/main/java/pages/vj/PassengerInformationPage.java:
```diff
@@ -0,0 +1,83 @@
+
+
+    @Step("Get reservation info")
+    public FlightInfo getReservationInfo(boolean isDeparture) {
```
avoid to use boolean in arguments. It's ambigious --> Use enum OR write higher method and covering all cases
Try to read below articles/blogs:
- http://dzone.com/articles/the-pitfalls-of-using-boolean-parameters-in-method
- https://alexkondov.com/should-you-pass-boolean-to-functions/

yenleagest replied:
Updated! Love your docs.

nbmhoang commented on src/main/java/pages/vj/PassengerInformationPage.java:
```diff
@@ -0,0 +1,83 @@
+        int adults = 0;
+        int children = 0;
+        int infants = 0;
+        if (info.length == 1) adults = Integer.parseInt(info[0].split(" ")[0]);
```
use switch-case please

yenleagest replied:
Done. Since this logic is used in two pages, I moved it to Passenger class.

nbmhoang commented on src/main/java/pages/vj/PassengerInformationPage.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Step("Get airport name")
+    private String getAirport(ElementsCollection reservationInfo, boolean isDeparture) {
```
same here, avoid use boolean in 2nd argument

yenleagest replied:
Done.

nbmhoang commented on src/main/java/pages/vj/PassengerInformationPage.java:
```diff
@@ -0,0 +1,83 @@
+    @Step("Get take off date")
+    private LocalDate getTakeOffDate(ElementsCollection reservationInfo) {
+        String dateText = reservationInfo.get(2).shouldNotHave(Condition.exactText("")).getText().split(",")[1].split("\\|")[0].trim();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
```
If this formatter is consistent across all languages, move it to a Constant class or make as a static variable in this page

yenleagest replied:
Done.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+
+
+    @Step("Close VJ advertisement")
+    public void closeAds() {
```
this method should place in VJ HomePage instead

yenleagest replied:
As discussed, it only shows in this page.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Step("Select cheapest flight")
+    public void selectCheapestFlight(boolean isReturnFlight) {
```
.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+        $(flightDetailsPanel).shouldBe(visible);
+    }
+
+    public void hitContinueButton(boolean isReturnFlight) {
```
.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+        waitForReservationInfoDisplayed(isReturnFlight);
+    }
+
+    private void waitForReservationInfoDisplayed(boolean isReturnFlight) {
```
.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+        int adults = 0;
+        int children = 0;
+        int infants = 0;
+        if (passengerInfo.length == 1)
```
same as above

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Step("Get airport name")
+    private String getAirport(ElementsCollection flightInfo, boolean isDeparture) {
```
.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+    @Step("Get take off date")
+    private LocalDate getTakeOffDate() {
+        String text = $(takeOffDate).shouldNotHave(Condition.exactText("")).getText().trim();
+        String baseUrl = System.getProperty("selenide.baseUrl");
```
same as above

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMMM d yyyy", Locale.ENGLISH);
+            return LocalDate.parse(cleanedWithYear, formatter);
+        } else if (baseUrl.endsWith("/vi")) {
+            Pattern pattern = Pattern.compile("(\\d{1,2})\\s+tháng\\s+(\\d{1,2})");
```
use datetimeformatter also, you could including 'tháng' in your pattern & use datetimeformatter builder for default year is current year

yenleagest replied:
Done. Thanks for calling it out, its a better approach.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+    private String getCheapestPrice() {
+        return $$(flightsPrice).shouldHave(CollectionCondition.sizeGreaterThan(0))
+                .stream()
+                .map(SelenideElement::getText) // get the price string like "4,090"
```
you could write the comparator inside the min( ) method, no need to clean the string then format it again

yenleagest replied:
Done. What a powerful interface.

nbmhoang commented on src/main/resources/locales/en.yml:
do we really need double quote " " to represent value as string?

yenleagest replied:
Removed, I confused that syntax with json lol.

nbmhoang commented on src/test/java/testcases/vj/SearchTicketsOnASpecificDayTest.java:
```diff
@@ -0,0 +1,63 @@
+    @Test(dataProvider = "dataByMethodLocale", groups = {"smoke", "regression"}, description = "Search and choose tickets on a specific day successfully")
+    public void searchTicketsOnASpecificDay(VJTestData data) {
+
+        departureDate = LocalDate.now().plusDays(1);
```
move to setUp

yenleagest replied:
Done

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,175 @@
+    private final By closeAdsBtn = By.cssSelector("[aria-label='close']");
+    private final By flightInfo = By.cssSelector(".MuiTypography-h5[variantmd='h3']");
+    private final By currency = By.cssSelector("p.MuiTypography-h4 ~ p.MuiTypography-body1");
+    private final By flightsPrice = By.cssSelector("p.MuiTypography-h4:not([variantlg='h3']");
```
Capturing this element for price comparison is acceptable. However, if you're extracting the original price, be aware that it may omit trailing zeros.

yenleagest replied:
Yah, Im aware of that. Thanks for your reminder.

catssaymeow committed:
Fix review comments

yenleagest commented:
Hi @nbmhoang Ive addressed all your comments. Please recheck when you have a chance.
I also attached the console logs as evidence that the test case is still working after the update 😎.
```
[INFO] Running TestSuite
14:58:55.086 [TestNG-test-VJRegression-1] INFO  com.codeborne.selenide.WebDriverThreadLocalContainer - No webdriver is bound to current thread: 34 - let's create a new webdriver
14:58:56.707 [TestNG-test-VJRegression-1] INFO  com.codeborne.selenide.webdriver.BrowserResizer - Set browser size to 1920x1080
14:58:56.974 [TestNG-test-VJRegression-1] INFO  com.codeborne.selenide.webdriver.WebDriverFactory - Selenide 7.9.3 / Selenium 4.33.0 / firefox 140.0.2 mac
14:59:11.795 [TestNG-test-VJRegression-1] INFO  pages.vj.SelectFlightOptionsPage - FlightInfo info: FlightInfo(currency=VND, departureAirport=Tp. Hồ Chí Minh, destinationAirport=Hà Nội, takeOffDate=2025-06-30, passenger=Passenger(adults=2, children=0, infants=0))
14:59:12.961 [TestNG-test-VJRegression-1] INFO  pages.vj.SelectFlightOptionsPage - FlightInfo info: FlightInfo(currency=VND, departureAirport=Hà Nội, destinationAirport=Tp. Hồ Chí Minh, takeOffDate=2025-07-03, passenger=Passenger(adults=2, children=0, infants=0))
14:59:14.118 [TestNG-test-VJRegression-1] INFO  pages.vj.PassengerInformationPage - Reservation info: FlightInfo(currency=VND, departureAirport=Tp. Hồ Chí Minh, destinationAirport=Hà Nội, takeOffDate=2025-06-30, passenger=Passenger(adults=2, children=0, infants=0))
14:59:14.176 [TestNG-test-VJRegression-1] INFO  pages.vj.PassengerInformationPage - Reservation info: FlightInfo(currency=VND, departureAirport=Hà Nội, destinationAirport=Tp. Hồ Chí Minh, takeOffDate=2025-07-03, passenger=Passenger(adults=2, children=0, infants=0))
14:59:14.704 [Dead threads watchdog] INFO  com.codeborne.selenide.impl.DeadThreadsWatchdog - Thread 34 is dead. Let's close its webdriver FirefoxDriver: firefox on mac (null).
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 20.523 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  23.065 s
[INFO] Finished at: 2025-06-29T14:59:14+07:00
[INFO] ------------------------------------------------------------------------
```

catssaymeow committed:
Fix issue with en locale

yenleagest commented:
I realized my changes broke the test when running with the en locale. Glad that I caught and fixed it before you reviewed.
![image](https://github.com/user-attachments/assets/437aa96a-bde2-4c4a-8276-a39619938a32)

nbmhoang commented on src/main/java/data/models/vj/Passenger.java:
```diff
@@ -0,0 +1,37 @@
+    private int infants;
+
+    @Step("Get passenger info")
+    public static Passenger getPassengerInfo(String text) {
```
Is it the constructor for this class, isn't it?

yenleagest replied:
No its not. This method take the text from the UI, parses and turns it to Passenger object. There are two pages use that logic so I moved it here to avoid duplicated codes.

yenleagest replied:
Did you mean we should move this logic into the constructor instead of having it as a static method?

yenleagest replied:
I don’t think it’s a good practice to turn this into a constructor. The current method name better demonstrates its purpose. Using a constructor would limit readability and flexibility.

nbmhoang replied:
You mean the name getPassengerInfo(String text) is enough to represent its logic? 😄 There's no UI text involved here — you're passing a raw string, extracting info from it, and then creating a new instance

nbmhoang replied:
You have an alternate option here: rename it, maybe extractInfo or fromString

yenleagest replied:
Yeah fromString sounds better.

nbmhoang commented on src/main/java/pages/vj/SelectFlightOptionsPage.java:
```diff
@@ -0,0 +1,136 @@
+        hitContinueButton(direction);
+    }
+
+    private void selectCheapestFlight(String price) {
```
this method should be "selectFlightByPrice"

yenleagest replied:
Updated.

nbmhoang reviewed (changes requested):
almost done

catssaymeow committed:
Rename methods

yenleagest commented:
> almost done

Please have another look.
